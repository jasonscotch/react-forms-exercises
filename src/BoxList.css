.BoxList {
  margin: 0 auto;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
}

.BoxList-div {
  margin-top: 50px;
  margin: 25px;
  font-weight: bold;
  display: flex;
  justify-content: center;
}

form label {
    margin: 10px;
}
form input {
    margin: 5px;
}
form button {
    margin: 5px;
}