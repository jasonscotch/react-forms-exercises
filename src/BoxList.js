import React, { useState } from "react";
import {v4 as uuid} from 'uuid';
import Box from "./Box";
import NewBoxForm from "./NewBoxForm";
import './BoxList.css'


const BoxList = () => {
    const INITIAL_STATE = [
        {id: uuid(), width:'250px', height:'250px', backgroundColor:'teal'},
        {id: uuid(), width:'250px', height:'250px', backgroundColor:'orange'}
    ]
    const [boxes, setBoxes] = useState(INITIAL_STATE);
    const addBox = (newBox) => {
        setBoxes(boxes => [...boxes, { ...newBox, id: uuid() }])
    }
    const removeBox = (idRemove) => {
        setBoxes(boxes => boxes.filter(box => box.id !== idRemove));
    } 

    return (
        <div className="BoxList">
            <h3>ALL THE BOXES!</h3>
            <NewBoxForm addBox={addBox} />
            <div className="BoxList-div">
                {boxes.map(({ id, width, height, backgroundColor }) => <Box id={id} width={width} height={height} backgroundColor={backgroundColor} key={id} removeBox={removeBox} />)}
            </div>
        </div>
    )
}

export default BoxList;