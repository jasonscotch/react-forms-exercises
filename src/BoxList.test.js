import React from "react";
import { render, fireEvent } from "@testing-library/react";
import BoxList from "./BoxList";

it("renders without crashing", function () {
  render(<BoxList />);
});

it("matches snapshot", function () {
  const { asFragment } = render(<BoxList />);
  expect(asFragment()).toMatchSnapshot();
});

it("should add a new box", () => {
    const { queryByText, getByLabelText, container } = render(<BoxList />);
    const widthInput = getByLabelText("Width:");
    const heightInput = getByLabelText("Height:");
    const colorInput = getByLabelText("Background Color:");
    const btn = queryByText("Add Box");
    expect(container).not.toHaveStyle('background-color: black');
    fireEvent.change(widthInput, { target: { value: '250px' } });
    fireEvent.change(heightInput, { target: { value: '250px' } });
    fireEvent.change(colorInput, { target: { value: 'black' } });
    fireEvent.click(btn);

    const lastBoxContainer = container.querySelectorAll('.Box-container:last-child')[0];
    const innerBox = lastBoxContainer.querySelector('div:first-child');
  
    expect(innerBox).toHaveStyle('width: 250px; height: 250px; background-color: black;');
});
  
it("should remove a box when X is clicked", () => {
    const { getByText, container } = render(<BoxList />);
    const lastBoxContainer = container.querySelectorAll('.Box-container:last-child')[0];
    const firstBoxContainer = container.querySelectorAll('.Box-container:first-child')[0];
    const lastInnerBox = lastBoxContainer.querySelector('div:first-child');
    const firstInnerBox = firstBoxContainer.querySelector('div:first-child');
    const xBtn = getByText('X', { selector: '.Box-container:nth-child(1) .X' });

    fireEvent.click(xBtn);
  
    expect(lastInnerBox).toHaveStyle('width: 250px; height: 250px; background-color: orange;');
    expect(firstInnerBox).not.toHaveStyle('width: 250px; height: 250px; background-color: orange;');
    
})
 
