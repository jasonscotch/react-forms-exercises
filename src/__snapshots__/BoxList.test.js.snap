// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="BoxList"
  >
    <h3>
      ALL THE BOXES!
    </h3>
    <form>
      <label
        for="width"
      >
        Width:
      </label>
      <input
        id="width"
        name="width"
        placeholder="Box Width"
        type="text"
        value=""
      />
      <label
        for="height"
      >
        Height:
      </label>
      <input
        id="height"
        name="height"
        placeholder="Box Height"
        type="text"
        value=""
      />
      <label
        for="backgroundColor"
      >
        Background Color:
      </label>
      <input
        id="backgroundColor"
        name="backgroundColor"
        placeholder="Box Background Color"
        type="text"
        value=""
      />
      <button>
        Add Box
      </button>
    </form>
    <div
      class="BoxList-div"
    >
      <div
        class="Box-container"
      >
        <div
          style="width: 250px; height: 250px; background-color: teal;"
        />
        <div
          class="X"
        >
          X
        </div>
      </div>
      <div
        class="Box-container"
      >
        <div
          style="width: 250px; height: 250px; background-color: orange;"
        />
        <div
          class="X"
        >
          X
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
